src/sumo_backend/utils.clj

```diff
@@ -0,0 +1,227 @@
+     (if (or
+          (= rank-keyword :maegashira)
+          (= rank-keyword :juryo))
+       (let [rank-number (get-rank-and-file-number rank-str)
+             maegashira-limit (get-lowest-maegashira-rank-in-tournament month year)]
+         (+ (rank-keyword ranks)
+            rank-number
+            (or (and (= rank-keyword :juryo) maegashira-limit) 0)))
+       (rank-keyword ranks)))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Get a Rikishi's rank from a bout 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn get-rank-in-bout
+  "for a given bout hashmap and rikishi name string, 
+   returns rikishi rank if it exists
+   else returns nil"
+  [rikishi bout]
+  (or
+   (and (= (:east bout) (clojure.string/upper-case rikishi)) (:east_rank bout))
+   (and (= (:west bout) (clojure.string/upper-case rikishi)) (:west_rank bout))
+   nil))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Get a Rikishi's opponent from a bout 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn get-bout-opponent
+  "for a given rikishi and bout
+   return the bout opponent name string"
+  [rikishi bout]
+  (or
+   (and (= (:east bout) (clojure.string/upper-case rikishi)) (:west bout))
+   (and (= (:west bout) (clojure.string/upper-case rikishi)) (:east bout))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Get a Rikishi's rank value from a bout 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn get-rank-value-in-bout
+  "given a rikishi and bout, return
+   rikishi's rank value"
+  [rikishi bout] ; "takakeisho" { :east "takakeiso" :rank "ozeki" ... } -> 2
+  (get-rank-value
+    (get-rank-in-bout rikishi bout)
+    (:month bout)
+    (:year bout)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Get Opponent's rank from a bout
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn get-opponent-rank-in-bout
+  "given a rikishi and bout, return
+   rikishi opponent's rank value"
+  [rikishi bout]
+  (get-rank-in-bout
+   (get-bout-opponent
+    rikishi
+    bout)
+   bout))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Get Opponent's rank value from a bout
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn get-opponent-rank-value-in-bout
+  "given a rikishi and bout, return
+   rikishi opponent's rank value"
+  [rikishi bout]
+  (get-rank-value-in-bout
+    (get-bout-opponent
+      rikishi
+      bout)
+    bout))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Did the rikishi win/lose given bout?
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn rikishi-win-or-lose-bout
+  "given a rikishi, bout, and an outcome
+   return true or false
+   according to the expression"
+  [rikishi outcome bout]
+  ((or 
+    (and (= outcome "lose") not=)
+    (and (= outcome "win") =))
+   (:winner bout) (clojure.string/upper-case rikishi)))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Get a Rikishi's rank for a tournament
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn get-rank-in-tournament
```

Comment:
This could be simplified as
```
(defn get-rank-in-tournament
  [rikishi bouts]
  (some
    #(get-rank-in-bout rikishi %)
    bouts))
```
`some` is equivalent to `(first (map ... ))` which, thanks to lazy evaluation, I think should be doing the same thing you've written.

Reply:
thanks @bslawski, I changed this function to be this way. When I first wrote a lot of this I was using the 
```
[first-thing & rest]
```

a lot and stepping through things one at a time. I found though as I went along that this is pretty slow. Would you write code with that `[first-thing & rest]` pattern or only if you couldn't accomplish what you wanted some other way first?